nestkernel/universal_data_logger.h

```diff
@@ -290,6 +302,273 @@ nest::UniversalDataLogger< HostNode >::DataLogger_::DataLogger_(
   recording_interval_ = req.get_recording_interval();
   recording_offset_ = req.get_recording_offset();
 }
+
+
+/**
+ * Universal data-logging plug-in for neuron models.
+ *
+ * This class provides logging of universal data such as
+ * membrane potentials or conductances in a way that is
+ * compatible with the DataLoggingRequest/DataLoggingReply
+ * read-out mechanism. It can handle logging multisynapse conductances,
+ * which was not possible with the previous implementation.
+ *
+ * The logger must be informed about any incoming DataLoggingRequest
+ * connections by calling connect_logging_device(). All incoming
+ * DataLoggingRequests should then be forwarded to the logger using
+ * handle().
+ *
+ * @note A reference to the host node is stored in the logger, for
+ *       access to the state and sending events. This requires a constructor
+ *       and a copy constructor for the HostNode::Buffers_, creating new
+ *       logger members initialized with the new node, e.g.
+ * @code
+ * struct Buffers_ {
+ *   Buffers_(aeif_cond_beta_multisynapse&);
+ *   Buffers_(const Buffers_&, aeif_cond_beta_multisynapse&);
+ *
+ *   DynamicUniversalDataLogger< aeif_cond_beta_multisynapse > logger_;
+ *
+ * ...
+ *
+ * nest::aeif_cond_beta_multisynapse::Buffers_::Buffers_(aeif_cond_beta_multisynapse&
+ * n)
+ * : logger_(n), ... {}
+ *
+ * nest::aeif_cond_beta_multisynapse::Buffers_::Buffers_(const Buffers_&,
+ * aeif_cond_beta_multisynapse& n)
+ * : logger_(n), ... {}
+ *
+ * nest::aeif_cond_beta_multisynapse::aeif_cond_beta_multisynapse()
+ * : ..., B_(*this) {}
+ *
+ * nest::aeif_cond_beta_multisynapse::aeif_cond_beta_multisynapse(const
+ * aeif_cond_beta_multisynapse& n)
+ * : ..., B_(n.B_, *this) {}
+ * @code
+ *
+ * @todo Could HostNode be passed as const& to handle() and record_data()?
+ *
+ * @note To avoid inclusion problems and code-bloat, the class
+ *       interface is defined in this file, while most of the
+ *       implementation is in the companion universal_data_logger_impl.h.
+ *       As a consequence, calls to UniversalDataLogger members should
+ *       only come from cpp files---do not inline them.
+ *
+ * @addtogroup Devices
+ */
+
+
+template < typename HostNode >
+class DynamicUniversalDataLogger
```

Comment:
This class including its internal classes is very closely related to `UniversalDataLogger`. Is it really necessary to implement the class completely from scratch? Can´t it inherit from `UniversalDataLogger` and save almost all the code in the process?

Reply:
This point is getting really complicated for me. Both classes are basically the same. The only difference is that one is constructed from a `RecordablesMap` and the other from a `DynamicRecordablesMap`, and most importantly in what is the type of contents the `node_access_` vector has. When using a `RecordablesMap`, the `node_access_` holds function pointers, and they should be called like `(*(node_access[j])()`. When using a  `DynamicRecordablesMap`, the `node_access_` holds `DataAccessFunctor`s, and they should be called like `node_access()`.
The problem with inheritance is that `DataLogger_` is private and all the important handling of `DataLogger_` should be rewritten for the derived class. I tried with template aliassing but for now without success.

Reply:
I spent some time trying to successfully share the code between `UniversalDataLogger` and `DynamicUniversalDataLogger`. I tried two approaches:
1. Make `DynamicUniversalDataLogger` a derived class of `UniversalDataLogger`. In principle this should allow us to use the base class' implementation of `handle`, `recorda_data`, `reset` and `init`. The problem is that all four methods use the member `data_loggers_`. The derived `DynamicUniversalDataLogger` defines its data member `data_loggers_` using its own `DataLogger_` private class, however this only hides the base class `data_loggers_` member. This leads to the problem that in base's `handle` method, the base's empty `data_loggers_` vector is used, instead of the populated derived class' `data_loggers_`. This leads to assertion errors and no recordings.
2. Add an extra template parameter `MapType` to `UniversalDataLogger` that distinguishes between the cases where `RecordablesMap` and `DynamicRecordablesMap` are used. This fails to compile because all neuron models assume that the `UniversalDataLogger` template has only one parameter, instead of two. I failed miserably at making a template specialization alias that forces the `MapType` parameter value. If someone with more experience could help with this, it would be great.

In this situation, I think that code duplication between the `UniversalDataLogger` and `DynamicUniversalDataLogger` classes is not so bad.

Reply:
@lucianopaz Thanks for exploring this! It seems then that we need to live with the code duplication here.